### PR TITLE
QFJ-973: Provide configuration to turn off checksum validation

### PR DESCRIPTION
**Pull request by WojciechZankowski**
Added:

* Setting and checksum validation flag
* Acceptance Test
* New entry in documentation

### CODE REVIEW

**Comment by chrjohn** [comment]
Hi @WojciechZankowski ,
thanks for the PR! :) I will add some comments later.
Cheers,
Chris

File: quickfixj-core/src/main/doc/usermanual/usage/configuration.html [hunk 1]
```diff
@@ -456,6 +456,14 @@ <H3>QuickFIX Settings</H3>
     <TD> Y<br>N</TD>
     <TD> Y </TD>
   </TR>
+  <TR ALIGN="left" VALIGN="middle">
+      <TD> <I>ProcessMessageWithInvalidChecksum</I> </TD>
```
Comment (chrjohn) [inline quickfixj-core/src/main/doc/usermanual/usage/configuration.html]: Just a gut feeling, but I think this setting should rather be named `ValidateChecksum`. Unfortunately the logic and default value needs to be the other way round then.
The flags and methods inside the Message class are also named `validateChecksum` already.
Maybe wait some days if one of the other regulars has another opinion on this. @philipwhiuk maybe?

File: quickfixj-core/src/main/java/quickfix/DefaultSessionFactory.java [hunk 2]
```diff
@@ -84,6 +84,14 @@ public Session create(SessionID sessionID, SessionSettings settings) throws Conf
             final boolean rejectGarbledMessage = getSetting(settings, sessionID,
                     Session.SETTING_REJECT_GARBLED_MESSAGE, false);
 
+            final boolean processMessageWithInvalidChecksum = getSetting(settings, sessionID,
+                    Session.SETTING_PROCESS_MESSAGE_WITH_INVALID_CHECKSUM, false);
+
+            if (rejectGarbledMessage && processMessageWithInvalidChecksum) {
+                throw new ConfigError("Not possible to reject garbled message and process " +
+                        "messages with invalid cheksum at the same time.");
```
Comment (chrjohn) [inline quickfixj-core/src/main/java/quickfix/DefaultSessionFactory.java]:
```suggestion
                        "messages with invalid checksum at the same time.");
```

File: quickfixj-core/src/main/doc/usermanual/usage/configuration.html (hunk 1 above)
Comment (WojciechZankowski) [reply quickfixj-core/src/main/doc/usermanual/usage/configuration.html]: Validate Checksum works for me as well. I will wait for opinion from others.

File: quickfixj-core/src/test/java/quickfix/DefaultSessionFactoryTest.java [hunk 3]
```diff
@@ -254,4 +259,13 @@ public void testTimestampPrecision() throws Exception {
         createSessionAndAssertConfigError("no exception", ".*No enum constant quickfix.UtcTimestampPrecision.PICOS.*");
     }
 
+    // QFJ-973
+    @Test
+    public void testRejectGarbledMessageAndProcessInvalidChecksumError() {
+        settings.setString(Session.SETTING_REJECT_GARBLED_MESSAGE, "Y");
+        settings.setString(Session.SETTING_PROCESS_MESSAGE_WITH_INVALID_CHECKSUM, "Y");
+        createSessionAndAssertConfigError("no exception", ".*Not possible to reject " +
+                "garbled message and process messages with invalid cheksum at the same time.*");
```
[the remark this answers is not in the document]
Comment (chrjohn) [reply quickfixj-core/src/test/java/quickfix/DefaultSessionFactoryTest.java]: Nevermind, just corrected it.

[the remark this answers is not in the document]
Comment (WojciechZankowski) [reply quickfixj-core/src/test/java/quickfix/DefaultSessionFactoryTest.java]: Thanks! Too much code coverage 😅

File: quickfixj-core/src/main/doc/usermanual/usage/configuration.html (hunk 1 above)
Comment (chrjohn) [reply quickfixj-core/src/main/doc/usermanual/usage/configuration.html]: I think you can go ahead.
Thanks

**Comment by WojciechZankowski** [comment]
I changed the name and I had to reverse the behavior in couple places. Rebased to newest master. Please take a look if I didn't miss anything, thanks!

File: quickfixj-core/src/main/java/quickfix/DefaultSessionFactory.java [hunk 4]
```diff
@@ -84,10 +84,10 @@ public Session create(SessionID sessionID, SessionSettings settings) throws Conf
             final boolean rejectGarbledMessage = getSetting(settings, sessionID,
                     Session.SETTING_REJECT_GARBLED_MESSAGE, false);
 
-            final boolean processMessageWithInvalidChecksum = getSetting(settings, sessionID,
-                    Session.SETTING_PROCESS_MESSAGE_WITH_INVALID_CHECKSUM, false);
+            final boolean validateChecksum = getSetting(settings, sessionID,
+                    Session.SETTING_VALIDATE_CHECKSUM, false);
```
Comment (chrjohn) [inline quickfixj-core/src/main/java/quickfix/DefaultSessionFactory.java]: I think this setting should have "true" as default.

Comment (chrjohn) [reply quickfixj-core/src/main/java/quickfix/DefaultSessionFactory.java]: This should hopefully fix the failing tests.

**Comment by philipwhiuk** [comment]
The name is fine. I can't say I'm likely to use this - we have test tools that auto-correct the checksum of manually copied FIX messages instead and I would not want it on in any real environment. 

I'm not sure I'm a huge fan of the constructor growth - it's getting to the point we have to consider a MessageBuilder. But that's a change we can put in QFJ 3.0

**Comment by chrjohn** [comment]
Hi @philipwhiuk ,
we sometimes have clients that face validation problems (e.g. due to special characters) and are happy when there is a setting that can be used until the real problem is solved or cause is found.
But in general I agree with you that it is better to always send messages with correct checksum.

Thanks,
Chris.

**Pull request merged**